### PR TITLE
[SuperTextField] - fix word jumping for expanded selection, and add tests (Resolves #560)

### LINKED ISSUE #560: Fix and test word jumping for SuperTextField
`AttributedTextEditingController`, which is used by `SuperTextField`, looks like it probably does the wrong thing when jumping by word at the extremities of the text.

Notice here that we always subtract at least `1` from the selection extent when jumping by word. At first glance, this would seem to give us an extent of `-1` if the user tries to jump a word when the caret sits at the beginning of the text.
https://github.com/superlistapp/super_editor/blob/main/super_editor/lib/src/infrastructure/super_textfield/infrastructure/attributed_text_editing_controller.dart#L712

Notice here that we always add at least `1` to the selection extent when jumping by word. At first glance, this would seem to push the caret beyond the end of the text, if the extent is already at the end.
https://github.com/superlistapp/super_editor/blob/main/super_editor/lib/src/infrastructure/super_textfield/infrastructure/attributed_text_editing_controller.dart#L757

Investigate these cases, add tests that demonstrate the behavior, and fix the implementation if it really is wrong.

### PR DESCRIPTION
**Pull request by matthew-carroll**
[SuperTextField] - fix word jumping for expanded selection, and add tests (Resolves #560)

- Fixed a word-jumping bug when selection is expanded
- Decomposed horizontal caret movement method to improve readability
- Added tests for word-jumping under various conditions.

### CODE REVIEW

File: super_editor/lib/src/infrastructure/super_textfield/infrastructure/attributed_text_editing_controller.dart [hunk 1]
```diff
@@ -691,79 +691,137 @@ class AttributedTextEditingController with ChangeNotifier {
...
+    }
+
+    final newExtent = text.text.moveOffsetUpstreamByCharacter(selection.extentOffset) ?? 0;
+    selection = TextSelection(
+      baseOffset: expandSelection ? selection.baseOffset : newExtent,
+      extentOffset: newExtent,
+    );
+  }
+
+  void _moveCaretDownstream({
```
Comment (angelosilvestre) [inline super_editor/lib/src/infrastructure/super_textfield/infrastructure/attributed_text_editing_controller.dart:759]: Here I think the comment should be "extent to the right side of the selection"

File: super_editor/test/super_textfield/attributed_text_editing_controller_test.dart [hunk 2]
```diff
@@ -1,15 +1,133 @@
...
+          textLayout: _NoOpTextLayout(),
+          expandSelection: false,
+          moveLeft: true,
+          movementModifier: MovementModifier.word,
+        );
 
-  group('AttributedTextEditingController', () {
-    group('user', () {
-      test('types hello **world** into empty field', () {
+        // Ensure that the selection didn't change.
```
Comment (angelosilvestre) [inline super_editor/test/super_textfield/attributed_text_editing_controller_test.dart:105]: The selection changes here

File: super_editor/test/super_textfield/attributed_text_editing_controller_test.dart [hunk 3]
```diff
@@ -1,15 +1,133 @@
...
+
+        // Move downstream by word.
+        controller.moveCaretHorizontally(
+          textLayout: _NoOpTextLayout(),
+          expandSelection: false,
+          moveLeft: false,
+          movementModifier: MovementModifier.word,
+        );
+
+        // Ensure that the selection didn't change.
```
Comment (angelosilvestre) [inline super_editor/test/super_textfield/attributed_text_editing_controller_test.dart:124]: Selection changes here

File: super_editor/lib/src/infrastructure/super_textfield/infrastructure/attributed_text_editing_controller.dart [hunk 4]
```diff
@@ -691,79 +691,137 @@ class AttributedTextEditingController with ChangeNotifier {
...
+    required ProseTextLayout textLayout,
+    required bool expandSelection,
+    required MovementModifier? movementModifier,
+  }) {
+    if (!selection.isCollapsed && !expandSelection) {
+      // The selection isn't collapsed and the user doesn't
+      // want to continue expanding the selection. Move the
+      // extent to the left side of the selection.
+      selection = TextSelection.collapsed(offset: selection.end);
```
Comment (angelosilvestre) [inline super_editor/lib/src/infrastructure/super_textfield/infrastructure/attributed_text_editing_controller.dart:768]: Here I think the comment should be "extent to the right side of the selection"

**Comment by matthew-carroll** [comment]
@angelosilvestre good catch on the comments. I updated them.

BTW, whenever you a review a PR, if you find that your approval is waiting on a few really minor details, feel free to approve and say something like "pending a few comments".

**Review by angelosilvestre — APPROVED** [review]
LGTM

**Pull request merged**